internal/metrics/remote_write.go

```diff
@@ -0,0 +1,394 @@
+package metrics
+
+import (
+	"context"
+	"encoding/json"
+	"io/ioutil"
+	"math"
+	"net/url"
+	"os"
+	"path"
+	"strconv"
+	"time"
+
+	"git.sr.ht/~spc/go-log"
+
+	"github.com/golang/snappy"
+	"github.com/project-flotta/flotta-operator/models"
+	config_util "github.com/prometheus/common/config"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/prompb"
+	"github.com/prometheus/prometheus/storage/remote"
+)
+
+const (
+	defaultWaitInterval         = 5 * time.Minute
+	defaultRequestNumSamples    = 10000
+	defaultRequestDuration      = time.Hour
+	defaultRequestRetryInterval = 10 * time.Second
+	LastWriteFileName           = "metrics-lastwrite"
+	DeviceLabel                 = "edgedeviceid"
+)
+
+type RemoteWrite struct {
+	deviceID             string
+	config               RemoteWriteConfiguration
+	tsdb                 API
+	LastWrite            time.Time
+	lastWriteFile        string
+	waitInterval         time.Duration
+	RangeDuration        time.Duration
+	RequestNumSamples    int
+	RequestRetryInterval time.Duration
+	Client               WriteClient
+}
+
+type RemoteWriteConfiguration struct {
+	Url     string // empty URL means that feature is disabled
+	Timeout time.Duration
+}
+
+func NewRemoteWrite(dataDir, deviceID string, tsdbInstance API) *RemoteWrite {
+	waitInterval, _ := time.ParseDuration(os.Getenv("REMOTEWRITE_WAIT_INTERVAL"))
+	if waitInterval == 0 {
+		waitInterval = defaultWaitInterval
+	}
+	RangeDuration, _ := time.ParseDuration(os.Getenv("REMOTEWRITE_RANGE_DURATION"))
+	if RangeDuration == 0 {
+		RangeDuration = defaultRequestDuration
+	}
+	requestNumSamples, _ := strconv.Atoi(os.Getenv("REMOTEWRITE_REQUEST_NUM_SAMPLES"))
+	if requestNumSamples == 0 {
+		requestNumSamples = defaultRequestNumSamples
+	}
+	requestRetryInterval, _ := time.ParseDuration(os.Getenv("REMOTEWRITE_REQUEST_RETRY_INTERVAL"))
+	if requestRetryInterval == 0 {
+		requestRetryInterval = defaultRequestRetryInterval
+	}
+
+	newRemoteWrite := RemoteWrite{
+		deviceID:             deviceID,
+		tsdb:                 tsdbInstance,
+		waitInterval:         waitInterval,
+		RangeDuration:        RangeDuration,
+		RequestNumSamples:    requestNumSamples,
+		RequestRetryInterval: requestRetryInterval,
+		lastWriteFile:        path.Join(dataDir, LastWriteFileName),
+	}
+
+	lastWriteBytes, err := ioutil.ReadFile(newRemoteWrite.lastWriteFile)
+	if err == nil {
+		if len(lastWriteBytes) != 0 {
+			i, err := strconv.ParseInt(string(lastWriteBytes), 10, 64)
+			if err == nil {
+				newRemoteWrite.LastWrite = time.Unix(i/1000000000, i%1000000000)
+			} else {
+				log.Error(err)
```

Comment:
The log line should have more data, like why it happens.

Reply:
we log the error. it has all the details. I do not know what the error is so I cannot add any relevant info.

Reply:
for me is a bit clear: `log.Error(err, "Device cannot parse last checkpoint value for remote writer")` than cannot convert atoi into int in the log that can comes from anywhere.

Reply:
@eloycoto done